### PR TITLE
Add support for Accton CSP7551

### PR DESCRIPTION
**Pull request by ideaship**
This PR adds code and configuration specific to the Accton CSP7551. Most notably, it adds platform drivers (cpld, sfp, fpga). It also contains a platform initialization script, creates configuration files for the network ports giving the CPU access to the Tofino chip, and increases the partition created for BISDN Linux from 6 GB to 160 GB.

### CODE REVIEW

File: conf/distro/bisdn-linux.conf [hunk 1]
```diff
@@ -18,7 +18,7 @@ FEEDNAMEPREFIX ??= "${DISTRO_VERSION}"
 FEEDURIPREFIX  ??= "nightly_builds/${MACHINE}/master/packages_latest-build"
 FEEDDOMAIN     ??= "http://repo.bisdn.de"
 
-DISTRO_FEATURES_DEFAULT ?= "acl argp ext2 ipv4 ipv6 largefile nfs pci pcmcia systemd usbgadget usbhost virtualization xattr"
+DISTRO_FEATURES_DEFAULT ?= "acl argp ext2 ipv4 ipv6 largefile nfs pci pcmcia seccomp systemd usbgadget usbhost virtualization xattr"
```
Comment (KanjiMonster) [inline conf/distro/bisdn-linux.conf:21]: This seems unrelated to the CSP support?

File: scripts/installer/install.sh [hunk 2]
```diff
@@ -315,7 +315,7 @@ onl_baseplatform="${onl_platform%-r*}"
 echo "BISDN Linux Installer"
 
 # part_size: BISDN Linux partition in MB
-part_size=6144
+part_size=${BISDN_LINUX_VOLUME_SIZE_MB:-6144}
```
Comment (KanjiMonster) [inline scripts/installer/install.sh:318]: Maybe move this to top where we also set `BISDN_LINUX_VOLUME_LABEL`, then it will be just overridden when including the platform.conf, and you don't need the if-not-set thingy here.

File: conf/distro/bisdn-linux.conf (hunk 1 above)
Comment (KanjiMonster) [reply conf/distro/bisdn-linux.conf:21]: Looking into this, this was changed with https://git.yoctoproject.org/meta-virtualization/commit/?h=kirkstone&id=51833d0e131155aeae563c08b892854265bd9a10, which only updated defaults, so seccomp is only a soft dependency not a hard one.

I suggest instead setting `PACKAGECONF` for docker-ce to the old value, to restore the old state.

**Review by KanjiMonster — CHANGES_REQUESTED** [review]

File: conf/distro/bisdn-linux.conf (hunk 1 above)
Comment (ideaship) [reply conf/distro/bisdn-linux.conf]: Addressed by https://github.com/bisdn/meta-switch/pull/131

**Comment by ideaship** [comment]
In response to the review for https://github.com/bisdn/meta-open-network-linux/pull/102, csp7551-specific changes have been moved to this PR.

File: recipes-kernel/accton-csp7551-cpld-mod/accton-csp7551-cpld-mod.bb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+SUMMARY = "CSP 7551 cpld module"
```
Comment (KanjiMonster) [inline recipes-kernel/accton-csp7551-cpld-mod/accton-csp7551-cpld-mod.bb:1]: If you have a publicly available source for them I would suggest directly checking them out from there instead of creating a local copy of them; makes updating them to a new version easier.

Having all modules built from one package is fine, they don't need to be separated.

Also having them in one package avoids the need for `KBUILD_MODPOST_WARN=1`, as modpost will be then able to resolve symbols between modules and set proper dependencies, and won't error out anymore.

File: scripts/installer/install.sh [hunk 4]
```diff
@@ -435,6 +441,8 @@ EOF
     ) > "$bisdn_linux_mnt/lib/systemd/network/90-enp.link"
 fi
 
+platform_setup
```
Comment (KanjiMonster) [inline scripts/installer/install.sh:444]: `platform_setup` should either come before all the automated/default setup or after, but not in between (both have their merits).

File: conf/machine/accton-csp7551.conf [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    ipmitool \
+"
+
+# add development tools
+BISDN_SWITCH_IMAGE_EXTRA_INSTALL:append = " \
+    fping \
+    iperf3 \
+    net-tools \
+    screen \
+    openvswitch \
```
[the remark this answers is not in the document]
Comment (ideaship) [reply conf/machine/accton-csp7551.conf]: No reason. Changed.

File: conf/machine/accton-csp7551.conf [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+"
+
+MACHINE_FEATURES += "pcbios efi"
+MACHINE_FEATURES:remove = "alsa"
+
+PLATFORM_VENDOR = "accton"
+BISDN_ONIE_MACHINE = "csp7551"
```
[the remark this answers is not in the document]
Comment (ideaship) [reply conf/machine/accton-csp7551.conf]: Removed.

File: conf/machine/accton-csp7551.conf [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+MACHINE_EXTRA_RDEPENDS += " \
+    accton-csp7551-cpld-mod \
+    accton-csp7551-fpga-mod \
+    accton-csp7551-sfp-mod \
+    platform-onl-init \
+"
+
+# Let platform init script load modules as needed
+KERNEL_MODULE_PROBECONF = "fpga_driver i2c-i801"
+module_conf_i2c-i801 = "blacklist i2c-i801"
```
[the remark this answers is not in the document]
Comment (ideaship) [reply conf/machine/accton-csp7551.conf]: It does. Changed.

File: recipes-kernel/accton-csp7551-cpld-mod/accton-csp7551-cpld-mod.bb (hunk 3 above)
Comment (ideaship) [reply recipes-kernel/accton-csp7551-cpld-mod/accton-csp7551-cpld-mod.bb:1]: The publicly available source does not look robust enough to pull from. I want to keep the source in this repo.

I agree that all modules could be built from one package, but there's no harm in having separate packages, and that's what we have done in meta-open-network-linux until we switched to building via ONL (which does not apply in this case because it is not an ONL platform).

File: scripts/installer/install.sh (hunk 4 above)
Comment (ideaship) [reply scripts/installer/install.sh]: Moved call down.

File: recipes-kernel/accton-csp7551-cpld-mod/accton-csp7551-cpld-mod.bb (hunk 3 above)
Comment (KanjiMonster) [reply recipes-kernel/accton-csp7551-cpld-mod/accton-csp7551-cpld-mod.bb:1]: How about https://github.com/WeichenChenACCTON/sonic-buildimage-wedge100bf-32qs/commit/bcf779f1798fbecc6024a05cbf4e994c820a67e2 ?

This one seems pretty stable, unmodified since august 2022, identical to the source you are using (at least the sonic-platform-modules-accton/ content is identical, including the modules), and is referenced in a [somewhat stale PR to sonic-buildimage](https://github.com/sonic-net/sonic-buildimage/pull/13099), which is where I found it.

The only difference to the source you are using is the sonic-buildimage commit it is based on, which does not matter for us at all.

And while we had modules as single packages in the past doesn't mean this is the right way to do it. Especially since it breaks inter-module dependencies.

So I would ask to at least merge them into a single package (e.g. accton-csp7551-modules or so), and preferably check out from git. Makes the source immediately obvious.

File: conf/machine/accton-csp7551.conf (hunk 7 above)
[the remark this answers is not in the document]
Comment (KanjiMonster) [reply conf/machine/accton-csp7551.conf]: There's actually no need to have this here, and this should rather be part of fpga_driver's recipe. Then you also shouldn't need an append.

File: recipes-kernel/accton-csp7551-cpld-mod/accton-csp7551-cpld-mod.bb (hunk 3 above)
Comment (ideaship) [reply recipes-kernel/accton-csp7551-cpld-mod/accton-csp7551-cpld-mod.bb]: Move all modules into a single package.

File: conf/machine/accton-csp7551.conf [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+"
+
+# add development tools
+MACHINE_EXTRA_RDEPENDS:append = " \
+    fping \
+    iperf3 \
+    net-tools \
+    screen \
+    openvswitch \
+"
```
[the remark this answers is not in the document]
Comment (ideaship) [reply conf/machine/accton-csp7551.conf]: Done.

File: recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+inherit module
+
+SRC_URI = " \
+          file://Makefile \
+          file://fpga_driver.c \
+          file://accton_i2c_cpld.c \
+          file://x86-64-accton-csp7551-sfp.c \
+          file://COPYING \
+          "
```
Comment (KanjiMonster) [inline recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb:13]: Please either do a base files + our modifications split as in https://github.com/bisdn/meta-ofdpa/tree/main/recipes-kernel/openbcm-gpl-modules or directly checkout from git (+ our modifications split).

Directly modifying the files makes it harder to do atomic updates of them in case accton releases new versions, and makes it harder to reapply our changes.

Also in case you use the bcm-gpl-modules approach, add a comment where you've taken the files from.

File: scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+EXTRA_CMDLINE_LINUX=""
+BISDN_LINUX_VOLUME_SIZE_MB=163840
+
+platform_setup() {
+    (cat <<EOF
+[Match]
+Path=pci-0000:18:00.0
+
+[Link]
+Description=Tofino ICE port (aka enp24s0)
```
Comment (KanjiMonster) [inline scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf:12]: `enp24s0` is systemd naming scheme dependent name, and can change with newer/older systemd versions.

Please choose a better identifier here (maybe there is a label on the box?)

Likewise all the other ones.

**Review by KanjiMonster — CHANGES_REQUESTED** [review]

File: conf/machine/accton-csp7551.conf [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+#@DESCRIPTION: Machine configuration for accton-csp7551 systems
+
+PREFERRED_PROVIDER_virtual/kernel ?= "linux-yocto-onl"
+PREFERRED_VERSION_linux-yocto-onl ?= "5.15%"
+
+KMACHINE:accton-csp7551 = "x86_64-intel"
+KERNEL_FEATURES += "\
+    bsp/x86_64-intel \
+"
```
[the remark this answers is not in the document]
Comment (ideaship) [reply conf/machine/accton-csp7551.conf:12]: Fixed.

File: conf/machine/accton-csp7551.conf [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+DEFAULTTUNE ?= "corei7-64"
+
+# https://git.yoctoproject.org/poky/tree/meta/conf/machine/include?h=kirkstone
+require conf/machine/include/x86/tune-corei7.inc
+require conf/machine/include/x86/x86-base.inc
+
+# https://github.com/bisdn/meta-open-network-linux/blob/master/conf/machine/include/onl.inc
```
[the remark this answers is not in the document]
Comment (ideaship) [reply conf/machine/accton-csp7551.conf]: Fixed.

File: conf/machine/accton-csp7551.conf [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    accton-csp7551-mods \
+    platform-onl-init \
+"
+
+# Firmware needed by qat and ice kernel modules (from poky's linux-firmware)
+MACHINE_EXTRA_RDEPENDS += " \
+    linux-firmware-qat \
+    linux-firmware-qat-license \
+    linux-firmware-ice \
+    linux-firmware-ice-license \
```
[the remark this answers is not in the document]
Comment (ideaship) [reply conf/machine/accton-csp7551.conf]: Done.

File: scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf (hunk 10 above)
Comment (ideaship) [reply scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf:12]: These are internal ports for communication between CPU and Tofino, so there are no labels. I thought the names given by systemd were still useful, but I have removed them now.

File: recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb (hunk 9 above)
Comment (ideaship) [reply recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb]: Done.

File: scripts/installer/install.sh [hunk 14]
```diff
@@ -5,6 +5,7 @@
 #  SPDX-License-Identifier:     GPL-2.0
 
 BISDN_LINUX_VOLUME_LABEL="BISDN-Linux"
+BISDN_LINUX_VOLUME_SIZE_MB=6144
```
Comment (KanjiMonster) [inline scripts/installer/install.sh:8]: I wonder if it wouldn't make more sense to instead of changing the default partition size to let the installer make the partition as large as possible. Then you don't need to set a custom size.

File: recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb (hunk 9 above)
Comment (KanjiMonster) [reply recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb]: > add a comment where you've taken the files from.

I meant that literally, add a comment to the recipe please ;P

File: recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb [hunk 15]
```diff
@@ -0,0 +1,19 @@
+SUMMARY = "CSP 7551 platform modules"
+LICENSE = "GPL-2.0-only"
```
Comment (KanjiMonster) [inline recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb:2]: Your COPYING, as well as `accton_i2c_cpld.c` and `x86-64-accton-csp7551-sfp.c` say:

>  This program is free software; you can redistribute it and/or modify
>  it under the terms of the GNU General Public License as published by
>  the Free Software Foundation; either version 2 of the License, or
>  (at your option) **any later** version.

The `fpga_driver.c` is missing a license header and only has a `MODULE_LICENSE("GPL");`, which can mean anything. but I guess we can assume they mean the same license.

**Review by KanjiMonster — CHANGES_REQUESTED** [review]

File: recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb (hunk 9 above)
Comment (ideaship) [reply recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb]: Done.

File: scripts/installer/install.sh (hunk 14 above)
Comment (ideaship) [reply scripts/installer/install.sh:8]: Certainly a possibility for another PR.

File: recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb (hunk 15 above)
Comment (ideaship) [reply recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb:2]: Considering that the driver is coming from a public SONiC repo, I believe that is a fair assumption to make.

Comment (KanjiMonster) [reply recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb:2]: I guess my point didn't come across: The actual license says GPL 2.0 or later, not GPL 2.0 only. The LICENSE in the recipe is wrong (and needs to be `GPL-2.0-or-later` or so).

File: scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+EXTRA_CMDLINE_LINUX=""
+BISDN_LINUX_VOLUME_SIZE_MB=163840
+
+platform_setup() {
+    (cat <<EOF
+[Match]
+Path=pci-0000:18:00.0
+
+[Link]
+Description=Tofino ICE port
```
Comment (KanjiMonster) [inline scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf:12]: Can we give these a bit better description? What exactly is this? What does ICE mean? What's the difference to the Tofino CPU port? And also maybe add a `#0` and `#1` or so, so they have unique descriptions.

**Review by KanjiMonster — CHANGES_REQUESTED** [review]

File: scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf (hunk 16 above)
Comment (ideaship) [reply scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf]: Not sure what you expect here. I amended the description.

File: recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb (hunk 15 above)
Comment (KanjiMonster) [reply recipes-kernel/accton-csp7551-mods/accton-csp7551-mods.bb:2]: The GPL license at the top of the file in `accton_i2c_cpld.c` explicitly says "either version 2 of the License, or (at your option) any later version", and the GPL license at the top `x86-64-accton-csp7551-sfp.c` says "either version 2 of the License, or (at your option) any later version". Both of these contain the "or later" wording, so they are by definition `GPL-2.0-or-later`.

See e.g. for comparison [x86-64-delta-ag7648-cpld-mux-1.c](https://github.com/opencomputeproject/OpenNetworkLinux/blob/master/packages/platforms/delta/x86-64/ag7648/modules/builds/x86-64-delta-ag7648-cpld-mux-1.c) which explicitly states it is version 2.

File: scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf (hunk 16 above)
Comment (KanjiMonster) [reply scripts/installer/machine/x86_64-accton_csp7551-r0/platform.conf]: I'm looking here for a clearer description on how these differ. You call all three of them CPU ports, but so are they functionally equivalent?

**Review by KanjiMonster — APPROVED** [review]
